dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java

```diff
@@ -157,25 +200,97 @@ public void deleteInterpretation( Interpretation interpretation )
         return interpretationStore.getAllOrderedLastUpdated( first, max );
     }
 
+    @Override
+    public Set<User> getMentionedUsers( String text ){
+        Set<User> users = new HashSet<>();
+        Matcher matcher = Pattern.compile( "(?:\\s|^)@([\\w+._-]+)" ).matcher( text );
+        while ( matcher.find() )
+        {
+            String username = matcher.group(1);
+            UserCredentials userCredentials = userService.getUserCredentialsByUsername( username );
+            if (userCredentials  != null )
+            {
+                users.add( userCredentials.getUserInfo() );
+            }
+        }
+        return users;
+    }
+    
+    @Override
+    public void sendNotifications( Interpretation interpretation, InterpretationComment comment,  Set<User> users)
+    {
+        if ( interpretation != null) {
+            String link = systemSettingManager.getInstanceBaseUrl();
+
+            switch ( interpretation.getType() )
+            {
+                case MAP:
+                    link += "/dhis-web-mapping/index.html?id=" + interpretation.getMap().getUid() + "&interpretationid=" + interpretation.getUid();
+                    break;
+                case REPORT_TABLE:
+                    link += "/dhis-web-pivot/index.html?id=" + interpretation.getReportTable().getUid() + "&interpretationid=" + interpretation.getUid();
+                    break;
+                case CHART:
+                    link += "/dhis-web-visualizer/index.html?id=" + interpretation.getChart().getUid() + "&interpretationid=" + interpretation.getUid();
+                    break;
+                case EVENT_REPORT:
+                    link += "/dhis-web-event-reports/index.html?id=" + interpretation.getChart().getUid() + "&interpretationid=" + interpretation.getUid();
+                    break;
+                case EVENT_CHART:
+                    link += "/dhis-web-event-visualizer/index.html?id=" + interpretation.getChart().getUid() + "&interpretationid=" + interpretation.getUid();
+                    break;
+                default:
```

Comment:
This will be fragile as it will break when we change or redeploy apps. Can we find a more robust way?

Reply:
@larshelge I have got the idea from [this bit of code](https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/pushanalysis/DefaultPushAnalysisService.java#L340-L360) which is already part of master branch. Do you want me to extract these two pieces of code into a single method or do you have anything more sophisticated in mind? Any suggestion? I also had a look at LocalAppStorageService but doesn't seem the way to go...